### ISSUE TITLE
Trivial anonymous functions cause "unbound type variable" IR typing errors

### ISSUE THREAD

**Issue by jstolarek** [links-lang/links#693]
code:
```
fun (x) {1}
```
error:
```
***: IR Type Error: Type variable 350 is unbound
occurring in IR binding:
(Ir.Let (
   (229,
    (`MetaTypeVar (Point @ 0x7f8165a5ba58 `Var ((350,
                                                 (CommonTypes.Linearity.Unl,
                                                  CommonTypes.Restriction.Any),
                                                 `Rigid))),
     "x", Var.Scope.Local)),
   ([], (Ir.Return (Ir.Variable 230)))))
```
config:
```
typecheck_ir=true
prelude=empty.links
print_types_pretty=false
optimise=false
debug=true
hide_fresh_type_vars=false
show_flavours=true
check_frontend_transformations=true
show_quantifiers=true
```
Other test cases that report same errors:
```
fun (x) {x}
fun (f) {fun (x){f (x)}}
fun (x) {<a>{x}</a>}
fun () {query {[]}}
```

**Comment by jstolarek** [comment]
Also, anonymous composition function causes the same error:

```
fun (f,g) {fun (x){f (g(x))}};
```

**Comment by jstolarek** [comment]
Also, disabling `fail_on_ir_type_error` reveals multiple unbound variable errors, e.g. for the anonymous composition function above we get:

```
Continuing after IR Type Error:
Type variable 353 is unbound
occurring in IR binding:
(Ir.Let ((231, (a::Any, "x", Var.Scope.Local)),
   ([], (Ir.Return (Ir.Variable 233)))))
backtrace:

-------------------------------------------------------------------


--------------------------------------------------------------------
Continuing after IR Type Error:
Type variable 354 is unbound
occurring in IR tail computation:
(Ir.Apply ((Ir.Project ("229", (Ir.Variable 239))), [(Ir.Variable 231)]))
backtrace:

-------------------------------------------------------------------


--------------------------------------------------------------------
Continuing after IR Type Error:
Type variable 354 is unbound
occurring in IR tail computation:
(Ir.Apply ((Ir.Project ("230", (Ir.Variable 239))), [(Ir.Variable 232)]))
backtrace:

-------------------------------------------------------------------

                                                                                                                                          [17/1848]
--------------------------------------------------------------------
Continuing after IR Type Error:
Type variable 353 is unbound
occurring in IR binding:
(Ir.Let ((229, ((a::Any) -b-> c::Any, "g", Var.Scope.Local)),
   ([], (Ir.Return (Ir.Variable 236)))))
backtrace:

-------------------------------------------------------------------


--------------------------------------------------------------------
Continuing after IR Type Error:
Type variable 355 is unbound
occurring in IR binding:
(Ir.Let ((230, ((a::Any) -b-> c::Any, "f", Var.Scope.Local)),
   ([], (Ir.Return (Ir.Variable 235)))))
backtrace:

-------------------------------------------------------------------

{}

--------------------------------------------------------------------
Continuing after IR Type Error:
Type variable 353 is unbound
occurring in IR value:
(Ir.Closure (234, [],
   (Ir.Extend ({229 => (Ir.Variable 229); 230 => (Ir.Variable 230); }, None))
   ))
backtrace:

-------------------------------------------------------------------
```

**Comment by jstolarek** [comment]
@frank-emrich, I made progress on this but I've run into what looks like bugs in
the IR typechecker and I need your help.

Firstly, an explanation of what is going on and why we get all these "unbound
type variable" errors.  Consider `fun (x) {1}` program.  Here's a post-frontend
AST:

```
Post-Frontend AST:
([],
 (Some (Block ([(Val (
  (Pattern.Variable ("it", forall a,b::Row.(a) -b-> Int)),
  ([a; a::Row],
   (Block ([(Fun
     fun_binder = ("_fun__g1", (a) -b-> Int);
     fun_linearity = CommonTypes.DeclaredLinearity.Unl;
     fun_definition =
     ([],
      ([[(Pattern.Variable ("x", a))]],
       (Block
          ([],
           (Constant
              (CommonTypes.Constant.Int 1))))));
     fun_location = CommonTypes.Location.Unknown;
     fun_signature = None; fun_unsafe_signature = false;
     fun_frozen = true }) ],
    (FreezeVar "_fun__g1")))),
  CommonTypes.Location.Unknown, None)],
  (FreezeVar "it")))))
```

The most outer `Block` is generated in `TypeSugar.type_check_general`:

```OCaml
block ~ppos
  ([with_pos ppos (Val ( variable_pat ~ppos ~ty:qtyp "it", (qs, body)
                       , loc_unknown, None))],
   freeze_var ~ppos "it")
```

Notice how the type of `"it"` is a quantified `forall a,b::Row.(a) -b-> Int` but
the inner type of `"_fun__g1"` is an unquantified `(a) -b-> Int`.  `"_fun__g1"` binder
is created in `desugarFuns`:

```
let e = block_node ([with_dummy_pos (
          Fun { fun_binder     = bndr
              ; fun_linearity  = lin
              ; fun_definition = def
              ; fun_location   = loc
              ; fun_signature  = None
              ; fun_frozen     = true
              ; fun_unsafe_signature = false; })],
         with_dummy_pos (FreezeVar f))
```

I attempted a naive fix of generalising the type of `bndr` at this point:

```
let (tyvars, _), gen_ty =
  Generalise.generalise (o#get_var_env ()) (Binder.to_type bndr) in
let gen_bndr = Binder.set_type bndr gen_ty in
let o = o#bind_binder gen_bndr in
let gen_bndr = Binder.set_type bndr gen_ty in
          Fun { fun_binder     = gen_bndr
              ; fun_linearity  = lin
              ; fun_definition = (tyvars, def)
              ; fun_location   = loc
              ; fun_signature  = None
              ; fun_frozen     = true
              ; fun_unsafe_signature = false; })],
                     with_dummy_pos (FreezeVar f))
```

I have no idea whether it is morally correct to run generalisation at this
point, but aside from this there seem to be some problems with the IR
typechecker.  Currently, when compiling `fun (x) {1}` it raises an exception:

```
Encountered exception during IR type-checking, saying:
(Invalid_argument List.fold_left2)occurring in IR binding:
(Ir.Let ((232, (forall a,b::Row.(a) -b-> Int, "it", Var.Scope.Global)),
   ([a; a::Row], (Ir.Return (Ir.Variable 231)))))
```

The reason is the line 1124 in `irCheck.ml`, which adds a forall to an actual
type:

```
let act_foralled = Types.for_all (tyvars, act) in
o#check_eq_types exp act_foralled (SBind orig);
```

Given that `act` was already generalised it leads to a nonsense actual type
`forall a,b::Row,a,b::Row.(a) -b-> Int` and an expected type `forall
a,b::Row.(a) -b-> Int`, which are obviously different. The exception itself is a result of passing lists of different length to `fold_left2` on line 305.  Changing the second line
to:

```
o#check_eq_types exp act (SBind orig);
```

makes the problem go away.  However, if I try to compile:

```
fun curry(f) {
  fun (x)(y) { f(x,y) }
}
```

I get an IR typing error:

```
Type mismatch:
 Expected:
  forall a::(Any,Any),b::(Any,Any),c::Row,d::(Any,Any),e::Row,f::Row.((a::Any, b::Any) -c-> d::Any) -e-> (a::Any) -f-> (b::Any) -c-> d::Any
 Actual:
  forall a::(Any,Any),b::(Any,Any),c::Row,d::(Any,Any),e::Row,f::Row.((a::Any, b::Any) -c-> d::Any) -e-> forall f::Row.(a::Any) -f-> (b::Any) -c-> d::Any
```

The actual type is definitely bogus, since it repeats the quantifier for `f`.
However, when I insert a `Debug.print` into `sugartoir` to see the actual type
assigned to a binding I get:

```
forall a::(Any,Any),b::(Any,Any),c::Row,d::(Any,Any),e::Row,f::Row.((a::Any, b::Any) -c-> d::Any) -e-> (a::Any) -f-> (b::Any) -c-> d::Any
```

which is correct and identical to expected.  Again, it seems that the IR
typechecker is doing a bad job at constructing the actual type.  Thoughts?

My code is on branch `js-T693-unbound-ir-tyvars` (c9cba3487c49e31d8ba062421120669ec5b0537d).

**Comment by frank-emrich** [comment]
I'm quite busy for the next days, but I should be able to have a look on the weekend

**Comment by jstolarek** [comment]
Sure, no problem.

**Comment by frank-emrich** [comment]
I've now had a look at this.

I'm on your branch `js-T693-unbound-ir-tyvars`, using the following settings:
```
typecheck_ir=true
prelude=empty.links
print_types_pretty=true
optimise=false
debug=true
hide_fresh_type_vars=false
show_flavours=true
check_frontend_transformations=true
show_quantifiers=true
show_post_frontend_ast=true
show_raw_type_vars=true
```

Then, the post-frontend AST looks like this:
```
 1([],
 2 (Some (Sugartypes.Block
 3          ([(Sugartypes.Val (
 4               (Sugartypes.Pattern.Variable
 5                  ("it", forall 350,351::Row.(350) -351-> Int)),
 6               ([350; 351::Row],
 7                (Sugartypes.Block
 8                   ([(Sugartypes.Fun
 9                        { Sugartypes.fun_binder =
10                          ("_fun__g1", forall 350,351::Row.(350) -351-> Int);
11                          fun_linearity = CommonTypes.DeclaredLinearity.Unl;
12                          fun_definition =
13                          ([350; 351::Row],
14                           ([[(Sugartypes.Pattern.Variable ("x", 350))]],
15                            (Sugartypes.Block
16                               ([],
17                                (Sugartypes.Constant
18                                   (CommonTypes.Constant.Int 1))))));
19                          fun_location = CommonTypes.Location.Unknown;
20                          fun_signature = None; fun_unsafe_signature = false;
21                          fun_frozen = true })
22                      ],
23                    (Sugartypes.FreezeVar "_fun__g1")))),
24               CommonTypes.Location.Unknown, None))
25             ],
26           (Sugartypes.FreezeVar "it")))))
```

This code looks wrong to me.
The function `_fun__g` defined on lines 8-22 has type `forall 350,351::Row.(350) -351-> Int`, which is also the type of the block defined on lines 7-23, because the block just returns the frozen version of `_fun__g`.

Due to the tyvars (= big lambdas) on line 6, the type of `it` would be the type of  `_fun__g`/the aforementioned block, generalized over 350, 351, yielding the overall type `forall 350,351::Row,350,351::Row.(350) -351-> Int`, which does not match the type `it` is annotated with.

In short, it looks like your patch has generalized the inner function definition of `_fun__g1`, but the big lambdas that generalize `it` are also still there, doing the same generalization again (using the same type variable names, which is probably problematic, too).

So it looks to me that the post-frontend AST that you stated in https://github.com/links-lang/links/issues/693#issuecomment-516863882 was actually well-typed, except that the generalization of the function happend on `it`, rather than the function itself.
However, that seems reasonable to me. In FreezeML, lambdas are not implicitly generalized, and we just happen to generalize top-level expressions if we can. Therefore, generalizing `it` rather than `_fun__g` seems like the right thing?

Next, I'm happy to have a look at why the original version of the example (the one you posted in https://github.com/links-lang/links/issues/693#issuecomment-516863882) fails to IR-typecheck.

**Comment by jstolarek** [comment]
Ok, I think I understand what is going on and what should be going on. Currently on master if we compile `f (x) {1}` the frontend generates:

```
let it : forall a,b::Row. a -b-> Int = /\a,b::Row. {
    let fun g1 (x : a) : a -b-> Int = 1
    in ~g1
  }
in ~it
```
Then in the IR `g1` is floated outside of `it`:
```
let fun g1 (y : a) : a -b-> Int = {
    let x : a = y in 1
} in
let it : forall a,b::Row. a -b-> Int = /\a,b::Row. g1 in
it
```
and now type variables `a` and `b` in the definition of `g1` are unbound.

In my patch I generalize `g1` so that the frontend produces:

```
let it : forall a,b::Row. a -b-> Int = /\a,b::Row. {
    let fun g1 (x : a) : forall a,b::Row. a -b-> Int = /\ a,b::Row. 1
    in ~g1
  }
in ~it
```
and then after floating `g1` in the IR we get:

```
let fun g1 (y : a) : forall a,b::Row. a -b-> Int = /\ a,b::Row. {
    let x : a = y in 1
} in
let it : forall a,b::Row. a -b-> Int = /\a,b::Row. g1 in
it
```
and now this is wrong because there are big lambdas in the bodies of both `it` and `g1`. My idea for a solution is to ensure that type variables are properly applied to `g1` when it is floated out:

```
let fun g1 (y : a) : forall a,b::Row. a -b-> Int = /\ a,b::Row. {
    let x : a = y in 1
} in
let it : forall a,b::Row. a -b-> Int = /\a,b::Row. (g1 a b) in
it
```
Does that look correct?

Yesterday during the dev meeting we discussed the possibility of first generalizing `g1` and only then generalizing `it`. However, I don't see an obvious way of doing that in the compilation pipeline. `it` is inserted and generalized during typechecking, while `g1` is generalized in `desugarFuns`, which is literally the last stage of the frontend.

**Comment by slindley** [comment]
Your solution looks sensible to me, modulo one small change (which you may or may not have already made in your actual patch): uses of variables in the final output must *all* be frozen.

    let fun g1 (y : a) : forall a,b::Row. a -b-> Int = /\ a,b::Row. {
        let x : a = ~y in 1
    } in
    let it : forall a,b::Row. a -b-> Int = /\a,b::Row. (~g1 a b) in
    ~it

I'm also a bit curious about why `y` needs to be rebound to `x` - I don't see why the body of `g1` isn't simply `1`.

Type-preserving flattening of let-bindings should really be done as part of the translation to the IR in `Sugartoir`. It may well be that we still hoist let-bindings without regard for type abstractions, but if that's the case then it is a bug.

**Comment by jstolarek** [comment]
> uses of variables in the final output must all be frozen.

I have confusingly used the same syntax for the frontend AST and the IR. So the snippets where the variables are not frozen are from the IR, where I believe there is no explicit freezing.

> I'm also a bit curious about why y needs to be rebound to x - I don't see why the body of g1 isn't simply 1.

I have no idea either, but that is what `sugartoir` produces.

**Comment by slindley** [comment]
Oh... in that case it all makes sense. Of course there's no freezing in the IR as it's a fully explicit language. The let binding is introduced by the pattern-matching compiler.

**Comment by jstolarek** [comment]
I just had a meeting with @frank-emrich and we wrote a fix. Apparently, my earlier fix for #534 was reverted during the FreezeML work and I just restored it. I still need to write the tests so no PR yet.

With my current patch we generate the code:
```
let it : forall a,b::Row. a -b-> Int = /\a,b::Row. {
    let fun g1 (x : a) : forall a,b::Row. a -b-> Int = /\ a,b::Row. 1
    in (~g1 a b)
  }
in ~it
```
@frank-emrich insists that we rename type variables in `g1` so that they are distinct from the ones used in `it`:
```
let it : forall a,b::Row. a -b-> Int = /\a,b::Row. {
    let fun g1 (x : c) : forall c,d::Row. c -d-> Int = /\ c,d::Row. 1
    in (~g1 a b)
  }
in ~it
```
However, we seem to be lacking facilities for doing this. While changing variable names in the signature of `g1` is fairly simple, we don't seem to have a way of changing types inside terms. More specifically, the signature on `x` needs to be changed from `(x : a)` to `(x : c)`. @frank-emrich wants to work on that though I am not convinced this belongs in a PR for this current bug.

**Comment by slindley** [comment]
I'm not sure where, but I know @SquidDev worked on code for renaming type variables in terms in the frontend. I guess here we're talking here about renaming those in IR terms. It could certainly be useful to have a facility to do that.

Currently we do not maintain the invariant that different type variables have distinct names - as is apparent from the correlation between `forall a` and `/\a` above. Note that Links *is not* FreezeML and there is no particular reason for these variables to match up. Also, given that we do not rely on uniqueness of type variables, there should be no need to implement the functionality you attribute to @frank-emrich as part of your patch. Having said that, there may be aesthetic and other reasons to consider generating code with distinct names for bound type variables.

**Comment by jstolarek** [comment]
@slindley: @frank-emrich just started a discussion on this in #734.

Very important note to self: current patch breaks code by reintroducing #678 into this example:

```
fun all(p, l) {
   switch (l) {
     case []    -> true
     case x::xs -> p(x) && all(p, xs)
   }
}

fun and(l) {
  all(fun(x) { x }, l)
}
```
Need to investigate.

**Comment by frank-emrich** [comment]
Given the discussion in #734, I'm fine with not uniquifying the type variables

**Comment by jamescheney** [comment]
I think (as @slindley says in #734) that it's good defensive programming practice to assume that there might be duplicate / shadowed bound variables in a given AST, unless there's a clearly documented convention or even an interface that enforces this invariant.

That said, implementing a variable uniqifier for the IR might also be good for "reverse fuzz testing": for all IR typing examples X, if X typechecks with unique variables but not with shadowed variables, or vice versa, then there is a bug somewhere.

**Comment by jstolarek** [comment]
Fixed in d72e68d156e439caa6440bb07ff67782492b4432
